presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java

```diff
@@ -158,4 +197,32 @@ public TaskContext addTaskContext(TaskStateMachine taskStateMachine, Session ses
         taskContexts.add(taskContext);
         return taskContext;
     }
+
+    private ListenableFuture<?> reserveMemoryUnsafe(long bytes)
+    {
+        if (willExceedMemoryLimit(bytes)) {
+            if (revocableFuture.isDone()) {
+                revocableFuture = SettableFuture.create();
+            }
+        }
+
+        ListenableFuture<?> future = memoryPool.reserve(queryId, bytes);
+
+        if (revocableFuture.isDone()) {
+            return future;
+        }
+        else {
+            return Futures.allAsList(ImmutableList.of(revocableFuture, future));
```

Comment:
what if there is no `revocableReserved` reserved? Then the returned future will never be done